tests/0107-regex_rebalance.c

```diff
@@ -0,0 +1,144 @@
+
+        if (rk == c1)
+                statep = &state1;
+        else if (rk == c2)
+                statep = &state2;
+        else
+                TEST_FAIL("Invalid rk %p", rk);
+
+        TEST_SAY("Rebalance for %s: %s:\n", rd_kafka_name(rk), rd_kafka_err2str(err));
+        test_print_partition_list(parts);
+
+        if (err == RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS)
+                rd_kafka_assign(rk, parts);
+        else if (err == RD_KAFKA_RESP_ERR__REVOKE_PARTITIONS)
+                rd_kafka_assign(rk, NULL);
+
+        *statep = err;
+}
+
+static void test_regex_rebalance(const char *assignor) {
+        const char *topic = test_mk_topic_name(__FUNCTION__ + 5, 1);
+        int64_t ts_start;
+        int wait_sec;
+        char newtopic[256];
+        rd_kafka_conf_t *conf;
+        char *topic_rgxa;
+        char *topic_rgxb;
+
+        state1 = state2 = 0;
+
+        rd_snprintf(newtopic, sizeof(newtopic), "%s.a", topic);
+
+        test_conf_init(&conf, NULL, 60);
+        test_conf_set(conf, "topic.metadata.refresh.interval.ms", "500");
+        test_conf_set(conf, "group.id", "rdkafkatest.consumer1");
+        test_conf_set(conf, "partition.assignment.strategy", assignor);
+
+        TEST_SAY("Creating 2 consumers\n");
+        c1 = test_create_consumer(topic, rebalance_cb, rd_kafka_conf_dup(conf), NULL);
+        c2 = test_create_consumer(topic, rebalance_cb, rd_kafka_conf_dup(conf), NULL);
+        rd_kafka_conf_destroy(conf);
+
+        TEST_SAY("Creating topic %s with 3 partitions\n", newtopic);
+        test_create_topic(c1, newtopic, 3, 1);
+
+        TEST_SAY("Subscribing\n");
+        // to force asymmetric subscriptions, cons1 = [ab], cons2 = [ac]
+        topic_rgxa = rd_strdup(tsprintf("^%s\\.[ab]", topic));
+        topic_rgxb = rd_strdup(tsprintf("^%s\\.[ac]", topic));
+        test_consumer_subscribe(c1, topic_rgxa);
+        test_consumer_subscribe(c2, topic_rgxb);
+
+
+        TEST_SAY("Waiting for initial assignment for both consumers\n");
+        while (state1 != RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS ||
+               state2 != RD_KAFKA_RESP_ERR__ASSIGN_PARTITIONS) {
+                test_consumer_poll_no_msgs("wait-rebalance", c1, 0, 1000);
+                test_consumer_poll_no_msgs("wait-rebalance", c2, 0, 1000);
+        }
+
+
+        TEST_SAY("Creating additional topics\n");
+        strcpy(newtopic, topic);
+        strcat(newtopic, ".b");
+        test_create_topic(c1, newtopic, 3, 1);
+        strcpy(newtopic, topic);
+        strcat(newtopic, ".c");
+        test_create_topic(c1, newtopic, 3, 1);
+
+        TEST_SAY("Wait 10 seconds for consumers not to crash\n");
+        wait_sec = test_quick ? 3 : 10;
+        ts_start = test_clock();
+        do {
+                test_consumer_poll_no_msgs("wait-stable", c1, 0, 1000);
+                test_consumer_poll_no_msgs("wait-stable", c2, 0, 1000);
+        } while (test_clock() < ts_start + (wait_sec * 1000000));
```

Comment:
Should wait for the assignment state (i.e., the asserts below) rather than a timeout.
Which also voids the need for a lower timeout in quick mode.